crates/ruma-macros/src/util.rs

```
@@ -324,29 +329,67 @@ impl StructFieldExt for Field {
+        self.attrs.iter().flat_map(AttributeExt::serde_meta_items)
     }
-}
 
-/// Whether the given attribute is a `#[serde(flatten)]` attribute.
-fn is_serde_flatten_attribute(attr: &Attribute) -> bool {
-    if !attr.path().is_ident("serde") {
-        return false;
+    fn has_serde_meta_item(&self, meta: SerdeMetaItem) -> bool {
+        self.serde_meta_items().any(|serde_meta| serde_meta == meta)
     }
+}
+
+/// Possible meta items for `#[serde(…)]` attributes.
+#[derive(Clone, Copy)]
+pub(crate) enum SerdeMetaItem {
+    /// `flatten`.
+    Flatten,
+
+    /// `default`.
+    Default,
+
+    /// `rename`.
+    Rename,
 
-    let mut contains_flatten = false;
-    let _ = attr.parse_nested_meta(|meta| {
-        if meta.path.is_ident("flatten") {
-            contains_flatten = true;
-            // Return an error to stop the parsing early.
-            return Err(meta.error("found"));
+    /// `alias`.
+    Alias,
+}
+
+impl SerdeMetaItem {
+    /// The string representation of this meta item.
+    fn as_str(self) -> &'static str {
+        match self {
+            Self::Flatten => "flatten",
+            Self::Default => "default",
+            Self::Rename => "rename",
+            Self::Alias => "alias",
         }
+    }
+}
 
-        Ok(())
-    });
+impl PartialEq<SerdeMetaItem> for syn::Meta {
+    fn eq(&self, other: &SerdeMetaItem) -> bool {
+        self.path().is_ident(other.as_str())
+    }
+}
 
-    contains_flatten
+/// Helper trait for a [`syn::Attribute`].
+pub(crate) trait AttributeExt {
+    /// Get the list of meta items if this is a `#[serde(…)]` attribute.
+    fn serde_meta_items(&self) -> impl Iterator<Item = syn::Meta>;
+}
+
+impl AttributeExt for Attribute {
+    fn serde_meta_items(&self) -> impl Iterator<Item = syn::Meta> {
+        if self.path().is_ident("serde")
+            && let syn::Meta::List(list) = &self.meta
+        {
+            list.parse_args_with(Punctuated::<syn::Meta, syn::Token![,]>::parse_terminated).ok()
```

This `ok` can swallow legit parsing errors, no?

---

Yes, but there shouldn't be a reason for parsing to fail if the attributes are correct. And if they are incorrect, serde should catch it.

---

Makes sense.